accepted/future-releases/inline-classes/feature-specification.md

```diff
@@ -0,0 +1,968 @@
+    if (localIt is num) {
+      yield localIt;
+    } else if (localIt is List<dynamic>) {
+      for (var element in localIt) {
+        yield* TinyJson(element).leaves;
+      }
+    } else {
+      throw "Unexpected object encountered in TinyJson value";
+    }
+  }
+}
+
+void main() {
+  var tiny = TinyJson(<dynamic>[<dynamic>[1, 2], 3, <dynamic>[]]);
+  print(tiny.leaves);
+  tiny.add("Hello!"); // Error.
+}
+```
+
+This is similar to the inline type in that it enforces the use of
+specific operations only (here we have just one: `leaves`), and it
+makes it an error to use instance methods of the representation (e.g.,
+`add`).
+
+Creation of wrapper objects consumes space and time. In the case where
+we wish to work on an entire data structure, we'd need to wrap each
+object as we navigate the data structure. For instance, we need to
+create a wrapper `TinyJson(element)` in order to invoke `leaves`
+recursively.
+
+In contrast, the inline class is zero-cost, in the sense that it does
+_not_ use a wrapper object, it enforces the desired discipline
+statically. In the inline class, the invocation of `TinyJson(element)`
+in the body of `leaves` can be eliminated entirely by inlining.
+
+Inline classes are static in nature, like extension members: an inline
+class declaration may declare some type parameters. The type
+parameters will be bound to types which are determined by the static
+type of the receiver. Similarly, members of an inline type are resolved
+statically, i.e., if `tiny.leaves` is an invocation of an inline getter
+`leaves`, then the declaration named `leaves` whose body is executed
+is determined at compile-time. There is no support for late binding of
+an inline member, and hence there is no notion of overriding. In return
+for this lack of expressive power, we get improved performance.
+
+
+## Syntax
+
+A rule for `<inlineClassDeclaration>` is added to the grammar, along
+with some rules for elements used in inline class declarations:
+
+```ebnf
+<inlineClassDeclaration> ::=
+  'inline' 'class' <typeIdentifier> <typeParameters>? <interfaces>?
+  '{'
+    (<metadata> <inlineMemberDeclaration>)*
+  '}'
+
+<inlineMemberDeclaration> ::= <classMemberDefinition>
+```
+
+*The token `inline` is not made a built-in identifier: the reserved
+word `class` that occurs right after `inline` serves to disambiguate
+the inline class declaration with a fixed lookahead.*
+
+A few errors can be detected immediately from the syntax:
+
+A compile-time error occurs unless the inline class declares exactly
+one instance variable. Let `v` be the name of said instance
+variable. The declaration of `v` must have the modifier `final`
+and it must not have the modifier `late`; otherwise a compile-time
+error occurs.
+
+The _name of the representation_ in an inline class declaration is the
+name `id` of the unique final instance variable that it declares, and
+the _type of the representation_ is the declared type of `id`.
```

Comment:
What about untyped instance fields? Do we disallow them or just type them as dynamic? Do we disallow instance fields with initializers?

Reply:
There are no untyped instance variables in Dart, at least not after type inference. So yes, we likely type them as `dynamic` or `Object?`, like we would any similar instance variable declaration in a normal `class`.

No need to disallow fields with initializers (and type inference for the variable should work as normal for that too).
Generally it should behave just like the similar `class`, except that you must have precisely one final instance variable, and can't declare the class abstract, and therefore cannot have abstract members.

Reply:
Having to infer the types of the field means that the representation type of the inline class will not be available until very late in the compilation process. This will give a complex dependency between inline class declarations and top level inference.

If the field doesn't have an initializer we can short-cut this by saying that the representation type is `dynamic`, which we would have otherwise inferred. For fields with initializers there is no shortcut and it could even lead to cyclic dependencies:
```
inline class A {
  final it = field;
}
var field = A().it;
```
So to me the question is, do we have a good use case for fields _with_ initializers that justifies the added complexity?

Reply:
I can see either saying "there can be no initializer", or "there must be an explicit type on the representation field".  I might lean a bit towards the latter.  Just say that we don't do field inference for inline class fields.

Reply:
> Just say that we don't do field inference for inline class fields.

Sounds good, done!